File: apps/ios/Wordocious/Sources/GamePersistence.swift
import Foundation
import WordociousCore

/// Persists in-progress games to disk so backgrounding/relaunch resumes the
/// exact board. Keyed by seed+mode, mirroring the web app's localStorage keys.
final class GamePersistence {
    static let shared = GamePersistence()
    private init() {}

    /// Save-format version, gating schema evolution the way the web's
    /// SAVE_VERSION does (use-game-snapshot.ts): a decodable-but-older save is
    /// DISCARDED on mismatch instead of silently restored into a newer
    /// reducer. Bump when GameState's persisted shape changes meaning.
    static let saveVersion = 1

    /// A game started close to local midnight can cross the day boundary.
    /// Keep yesterday's IN-PROGRESS daily loadable for this window instead of
    /// wiping it at the first launch after midnight (web
    /// DAILY_CROSS_MIDNIGHT_GRACE_MS parity — 4h).
    static let crossMidnightGraceSeconds: TimeInterval = 4 * 60 * 60

    /// Versioned on-disk envelope. Legacy files are a bare GameState — decoded
    /// as v1 by the fallback in load(), so shipping this doesn't wipe anyone's
    /// in-flight game.
    private struct VersionedSave: Codable {
        let version: Int
        let state: GameState
    }

    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private var directory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent("games", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private func key(seed: String, mode: GameMode) -> String {
        let safeSeed = seed.replacingOccurrences(of: "/", with: "_")
        return "wordocious-\(mode.rawValue)-\(safeSeed)"
    }

    private func url(seed: String, mode: GameMode) -> URL {
        directory.appendingPathComponent("\(key(seed: seed, mode: mode)).json")
    }

    func save(_ state: GameState) {
        let envelope = VersionedSave(version: Self.saveVersion, state: state)
        guard let data = try? encoder.encode(envelope) else { return }
        try? data.write(to: url(seed: state.seed, mode: state.mode), options: .atomic)
    }

    func load(seed: String, mode: GameMode) -> GameState? {
        let u = url(seed: seed, mode: mode)
        guard let data = try? Data(contentsOf: u) else { return nil }
        if let envelope = try? decoder.decode(VersionedSave.self, from: data) {
            guard envelope.version == Self.saveVersion else {
                // Versioned but older/newer: discard rather than restore a
                // shape this reducer no longer means the same thing by.
                try? FileManager.default.removeItem(at: u)
                return nil
            }
            return envelope.state
        }
        // Legacy pre-envelope file (bare GameState) — today's shape IS v1.
        return try? decoder.decode(GameState.self, from: data)
    }

    /// If today's daily has no save but YESTERDAY's daily for this mode is
    /// still in progress and was last saved within the cross-midnight grace
    /// window, return yesterday's seed so the caller can resume it (it records
    /// to yesterday — the day always derives from the seed, never the clock).
    /// Web parity: use-game-snapshot.ts's stillPlaying + withinGrace check.
    func gracedYesterdaySeed(todaySeed: String, mode: GameMode) -> String? {
        guard let todayStr = getDailySeedDate(todaySeed),
              let today = dateFromDayString(todayStr),
              let yesterday = Calendar.current.date(byAdding: .day, value: -1, to: today)
        else { return nil }
        let yesterdayStr = dayString(from: yesterday)
        let seed = generateDailySeed(date: yesterdayStr, gameMode: mode.rawValue)
        let u = url(seed: seed, mode: mode)
        guard let attrs = try? FileManager.default.attributesOfItem(atPath: u.path),
              let modified = attrs[.modificationDate] as? Date,
              Date().timeIntervalSince(modified) < Self.crossMidnightGraceSeconds,
              let state = load(seed: seed, mode: mode),
              state.status == .playing
        else { return nil }
        return seed
    }

    private func dayString(from date: Date) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = Calendar(identifier: .gregorian)
        f.timeZone = .current
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: date)
    }

    private func dateFromDayString(_ s: String) -> Date? {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = Calendar(identifier: .gregorian)
        f.timeZone = .current
        f.dateFormat = "yyyy-MM-dd"
        return f.date(from: s)
    }

    func clear(seed: String, mode: GameMode) {
        try? FileManager.default.removeItem(at: url(seed: seed, mode: mode))
    }

    /// Wipe ALL on-device game saves (every mode/seed) plus the elapsed/hint
    /// keys. Called on sign-out so the next session — a guest, or a different
    /// account — never inherits the previous user's in-progress or finished
    /// boards (which the completed-daily card reads from local persistence).
    func clearAllSaves() {
        let fm = FileManager.default
        if let files = try? fm.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil) {
            for f in files { try? fm.removeItem(at: f) }
        }
        let defaults = UserDefaults.standard
        for key in defaults.dictionaryRepresentation().keys
        where key.hasPrefix("wordocious-elapsed-") || key.hasPrefix("wordocious-hints-")
            || key.hasPrefix("pn-save-") {
            defaults.removeObject(forKey: key)
        }
    }

    // MARK: - Active-play elapsed (milliseconds), persisted per seed+mode

    private func elapsedKey(seed: String, mode: GameMode) -> String {
        "wordocious-elapsed-\(mode.rawValue)-\(seed.replacingOccurrences(of: "/", with: "_"))"
    }

    func saveElapsed(_ ms: Double, seed: String, mode: GameMode) {
        UserDefaults.standard.set(ms, forKey: elapsedKey(seed: seed, mode: mode))
    }

    func loadElapsed(seed: String, mode: GameMode) -> Double {
        UserDefaults.standard.double(forKey: elapsedKey(seed: seed, mode: mode))
    }

    /// Storage hygiene: daily games are keyed by date, so their save files (and
    /// elapsed/hint keys) accumulate forever. Remove any daily entry for a date
    /// other than today. Call once at launch. (Web has no such buildup — it uses
    /// a single per-mode localStorage key.)
    func cleanupStaleDailyGames() {
        let today = LeaderboardService.todayLocal()
        // Yesterday's saves get the cross-midnight grace: an IN-PROGRESS daily
        // last touched within the window survives the sweep so
        // gracedYesterdaySeed can resume it (web parity — this sweep used to
        // wipe a 11:58pm game at the 12:05am relaunch).
        let yesterday = Calendar.current.date(byAdding: .day, value: -1, to: Date()).map(dayString(from:))
        var keptDates: Set<String> = [today]
        let fm = FileManager.default
        if let files = try? fm.contentsOfDirectory(at: directory, includingPropertiesForKeys: [.contentModificationDateKey]) {
            for f in files {
                let base = f.lastPathComponent.replacingOccurrences(of: ".json", with: "")
                guard let r = base.range(of: "-daily-") else { continue }
                let date = String(base[r.upperBound...].prefix(10))   // YYYY-MM-DD
                guard date.count == 10, date != today else { continue }
                if date == yesterday,
                   let attrs = try? fm.attributesOfItem(atPath: f.path),
                   let modified = attrs[.modificationDate] as? Date,
                   Date().timeIntervalSince(modified) < Self.crossMidnightGraceSeconds,
                   let data = try? Data(contentsOf: f),
                   let state = (try? decoder.decode(VersionedSave.self, from: data))?.state
                       ?? (try? decoder.decode(GameState.self, from: data)),
                   state.status == .playing {
                    keptDates.insert(date)
                    continue
                }
                try? fm.removeItem(at: f)
            }
        }
        let defaults = UserDefaults.standard
        for key in defaults.dictionaryRepresentation().keys
        where key.contains("-daily-") && (key.hasPrefix("wordocious-elapsed-") || key.hasPrefix("wordocious-hints-")) {
            if let r = key.range(of: "-daily-") {
                let date = String(key[r.upperBound...].prefix(10))
                if date.count == 10, !keptDates.contains(date) { defaults.removeObject(forKey: key) }
            }
        }
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/GamePersistence.kt
package com.wordocious.app.data

import android.content.Context
import com.wordocious.app.App
import com.wordocious.core.GameMode
import com.wordocious.core.GameState
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Local game persistence — mirrors the iOS local-resume behavior (UserDefaults).
 * Stores the full GameState (JSON) keyed by seed+mode so a daily/in-progress game
 * resumes after the app is backgrounded or closed. The finished state is kept too,
 * so re-opening a completed daily lands on the post-game screen (web parity).
 */
object GamePersistence {
    /** Save-format version, gating schema evolution the way the web's
     *  SAVE_VERSION does (use-game-snapshot.ts): a decodable-but-older save is
     *  DISCARDED on mismatch instead of silently restored into a newer
     *  reducer (ignoreUnknownKeys made that especially easy here). Bump when
     *  GameState's persisted shape changes meaning. */
    const val SAVE_VERSION = 1

    /** A game started close to local midnight can cross the day boundary.
     *  Keep yesterday's IN-PROGRESS daily loadable for this window (web
     *  DAILY_CROSS_MIDNIGHT_GRACE_MS parity — 4h). */
    const val CROSS_MIDNIGHT_GRACE_MS = 4L * 60 * 60 * 1000

    /** Versioned envelope. Legacy entries are a bare GameState — decoded as v1
     *  by the fallback in [decodeVersionedSave], so shipping this doesn't wipe
     *  anyone's in-flight game. savedAt powers the cross-midnight grace
     *  (SharedPreferences has no file mtime). */
    @Serializable
    data class VersionedSave(val version: Int, val savedAt: Long, val state: GameState)

    private val prefs by lazy {
        App.instance.getSharedPreferences("wordocious_games", Context.MODE_PRIVATE)
    }
    private val json = Json { ignoreUnknownKeys = true }

    private fun key(seed: String, mode: GameMode) = "game-${mode.name}-$seed"

    // Pure encode/decode, extracted so the version-gate is unit-testable
    // (GamePersistenceTest) without SharedPreferences.
    fun encodeVersionedSave(state: GameState, savedAt: Long = System.currentTimeMillis()): String =
        json.encodeToString(VersionedSave.serializer(), VersionedSave(SAVE_VERSION, savedAt, state))

    /** Decode a stored entry: envelope with matching version, legacy bare
     *  GameState (treated as v1, savedAt unknown → 0), or null (mismatched
     *  version / corrupt). */
    fun decodeVersionedSave(raw: String): VersionedSave? {
        runCatching { json.decodeFromString(VersionedSave.serializer(), raw) }.getOrNull()?.let {
            return if (it.version == SAVE_VERSION) it else null
        }
        return runCatching { json.decodeFromString(GameState.serializer(), raw) }.getOrNull()
            ?.let { VersionedSave(SAVE_VERSION, 0L, it) }
    }

    fun save(seed: String, mode: GameMode, state: GameState) {
        runCatching {
            prefs.edit().putString(key(seed, mode), encodeVersionedSave(state)).apply()
        }
    }

    fun load(seed: String, mode: GameMode): GameState? =
        prefs.getString(key(seed, mode), null)?.let { decodeVersionedSave(it)?.state }

    /** If today's daily has no save but YESTERDAY's daily for this mode is
     *  still in progress and was saved within the grace window, return
     *  yesterday's seed so the caller can resume it (it records to yesterday —
     *  the day always derives from the seed, never the clock). Mirrors iOS
     *  GamePersistence.gracedYesterdaySeed / web use-game-snapshot.ts. */
    fun gracedYesterdaySeed(todaySeed: String, mode: GameMode): String? {
        val todayStr = com.wordocious.core.getDailySeedDate(todaySeed) ?: return null
        val yesterdayStr = previousDay(todayStr) ?: return null
        val seed = com.wordocious.core.generateDailySeed(yesterdayStr, mode.name)
        val raw = prefs.getString(key(seed, mode), null) ?: return null
        val saved = decodeVersionedSave(raw) ?: return null
        val fresh = System.currentTimeMillis() - saved.savedAt < CROSS_MIDNIGHT_GRACE_MS
        return if (fresh && saved.state.status == com.wordocious.core.GameStatus.PLAYING) seed else null
    }

    /** The `YYYY-MM-DD` after "-daily-" in a persistence key, or null for
     *  practice/VS keys. Pure — unit-tested. */
    fun dailyDateFrom(key: String): String? {
        val i = key.indexOf("-daily-")
        if (i < 0) return null
        val date = key.drop(i + "-daily-".length).take(10)
        return if (Regex("""\d{4}-\d{2}-\d{2}""").matches(date)) date else null
    }

    private fun previousDay(day: String): String? = runCatching {
        val f = java.text.SimpleDateFormat("yyyy-MM-dd", java.util.Locale.US)
        f.timeZone = java.util.TimeZone.getDefault()
        val d = f.parse(day) ?: return null
        java.util.Calendar.getInstance().apply { time = d; add(java.util.Calendar.DAY_OF_YEAR, -1) }
            .let { f.format(it.time) }
    }.getOrNull()

    /** Storage hygiene (iOS cleanupStaleDailyGames parity — Android previously
     *  had NO sweep, so per-seed daily saves accumulated forever). Removes
     *  game/elapsed/hint entries for any daily date other than today, except
     *  yesterday's still-in-progress game within the cross-midnight grace.
     *  Call once at app start. */
    fun cleanupStaleDailyGames(today: String) {
        runCatching {
            val yesterday = previousDay(today)
            val all = prefs.all
            // Pass 1 — decide which dates survive, from game- entries only:
            // today always; yesterday iff some game is still PLAYING within
            // the grace window. (Two passes so an elapsed/hint key seen before
            // its game- key can't be swept out from under a graced game.)
            val keep = mutableSetOf(today)
            if (yesterday != null) {
                val graced = all.any { (k, v) ->
                    k.startsWith("game-") && dailyDateFrom(k) == yesterday && v is String &&
                        decodeVersionedSave(v)?.let {
                            it.state.status == com.wordocious.core.GameStatus.PLAYING &&
                                System.currentTimeMillis() - it.savedAt < CROSS_MIDNIGHT_GRACE_MS
                        } == true
                }
                if (graced) keep.add(yesterday)
            }
            // Pass 2 — remove every daily-keyed entry whose date didn't survive.
            val editor = prefs.edit()
            for ((k, _) in all) {
                val date = dailyDateFrom(k) ?: continue
                if (date !in keep) editor.remove(k)
            }
            editor.apply()
        }
    }

    fun clear(seed: String, mode: GameMode) {
        prefs.edit().remove(key(seed, mode)).apply()
    }

    /** Wipe ALL on-device game saves (every mode/seed, plus elapsed/hint keys).
     *  Called on sign-out so the next session never inherits the previous user's
     *  boards — the completed-daily card reads from this local store. */
    fun clearAll() {
        prefs.edit().clear().apply()
    }

    // Elapsed-at-finish: frozen when the game ends so a re-opened finished game
    // shows the recorded time (not a value that keeps growing from startTime).
    private fun elapsedKey(seed: String, mode: GameMode) = "elapsed-${mode.name}-$seed"

    fun saveElapsed(seed: String, mode: GameMode, seconds: Int) {
        prefs.edit().putInt(elapsedKey(seed, mode), seconds).apply()
    }

    fun loadElapsed(seed: String, mode: GameMode): Int? {
        val k = elapsedKey(seed, mode)
        return if (prefs.contains(k)) prefs.getInt(k, 0) else null
    }

    // Hint UI state (Six/Seven/ProperNoundle) — the board guesses already persist
    // via save(), but the revealed clue/vowel/consonant flags + clue text live
    // outside GameState, so they need their own slot to survive a resume
    // (mirrors iOS persistHintUI/restoreHintUI).
    @Serializable
    data class HintState(
        val clue: String? = null,
        val vowelRevealed: String? = null,
        val consonantRevealed: String? = null,
        val vowelUsed: Boolean = false,
        val consonantUsed: Boolean = false,
    )

    private fun hintKey(seed: String, mode: GameMode) = "hints-${mode.name}-$seed"

    fun saveHints(seed: String, mode: GameMode, state: HintState) {
        runCatching {
            prefs.edit().putString(hintKey(seed, mode), json.encodeToString(HintState.serializer(), state)).apply()
        }
    }

    fun loadHints(seed: String, mode: GameMode): HintState? =
        prefs.getString(hintKey(seed, mode), null)?.let {
            runCatching { json.decodeFromString(HintState.serializer(), it) }.getOrNull()
        }
}
